sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,488 @@
+/// Access must be guarded by the \ref Sync::getGlobalLock()
+std::unordered_map<SymbolSetId,
+                   std::unique_ptr<std::vector<RTDeviceBinaryImageUPtr>>>
+    m_DeviceImages;
+
+using StrToKSIdMap = std::unordered_map<string_class, SymbolSetId>;
+/// Maps names of kernels from a specific OS module (.exe .dll) to their set
+/// id (the sets are disjoint).
+std::unordered_map<OSModuleHandle, SymbolSetId> m_SymbolSets;
+```
+Assume each device image represents some combination of symbols and different
+device images may contain only exactly the same or not overlapping combination
+of symbols. If it is not so, there can be two cases:
+  - Symbols are the same. In this case it doesn't matter which device image is
+  taken to use duplicated symbol
+  - Symbols are not the same. In this case ODR violation takes place, such
+  situation leads to undefined behaviour. For more details refer to
+  [ODR violations](#ODR-violations) section.
+
+Each combination of symbols is assigned with an Id number - symbol set Id.
+A combination of symbols can exist in different formats (i.e. SPIR-V/AOT
+compiled binary and etc).
+`m_DeviceImages` maps an Id number to an array with device images which represent
+the same combination of symbols in different formats.
+`m_SymbolSets` contains mapping from symbol name to symbol set Id for each OS
+module (.exe/.so/.dll).
+`std::unordered_map` allows to search and access its elements with constant-time
+complexity.
+
+Before compilation of device image, to execute a kernel RT checks if the image
+contains any import information in its properties and if it does, then RT
+performs device images collection in order to resolve dependencies.
+
+Ids of all needed symbol sets are found by iterating through
+`m_SymbolSets` map, i.e. iterating through all available OS modules without
+predefined order and searching for first unresolved symbol in list of imports
+of target device image. Once device image that contains first symbol is
+met, remaining exported symbols are checked in found image. If
+they match some imported symbols then these matched symbols will be marked as
+resolved. The procedure repeats until all imported symbols are marked as
+resolved.
+
+#### Program caching
+
+Existing support for device code caching can be re-used to cache
+dynamically linked programs with slight changes.
+
+##### In-memory cache
+
+The existing mechanism of caching can be re-used in presence of dynamic
+linking. Example:
+```
+// Application
+SYCL_EXTERNAL void LibFunc1();
+SYCL_EXTERNAL void LibFunc2();
+
+Q.submit([&](cl::sycl::handler &CGH) {
+CGH.parallel_for<InternalKernel>( ... )
+}); // 1. Device Image is compiled and linked into a program and saved in cache
+    // 2. Prepared program is used to enqueue kernel
+
+Q.submit([&](cl::sycl::handler &CGH) {
+handler.parallel_for([] { LibFunc1(); }); // Prepared program is used to enqueue kernel
+});
+
+// Library
+SYCL_EXTERNAL void LibFunc1() {
+// ...
+}
+
```

Comment:
```suggestion
SYCL_EXTERNAL void LibFunc2() {
// ...
}
```

Reply:
This one is not used in the example, so I'll just probably remove the declaration from the app.